client/rest-high-level/src/main/java/org/elasticsearch/client/security/support/expressiondsl/fields/FieldRoleMapperExpression.java

```diff
@@ -0,0 +1,121 @@
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * An expression that evaluates to <code>true</code> if a field (map element) matches
+ * the provided values. A <em>field</em> expression may have more than one provided value, in which
+ * case the expression is true if <em>any</em> of the values are matched.
+ * <p>
+ * Expression builder example:
+ * <pre>
+ * {@code
+ * final RoleMapperExpression usernameExpression = FieldRoleMapperExpression.ofUsername("user1@example.org");
+ * }
+ * </pre>
+ */
+public class FieldRoleMapperExpression implements RoleMapperExpression {
+    private final String NAME = "field";
+
+    private final String field;
+    private final List<Object> values;
+
+    public FieldRoleMapperExpression(final String field, final Object... values) {
+        if (field == null || field.isEmpty()) {
+            throw new IllegalArgumentException("null or empty field name (" + field + ")");
+        }
+        if (values == null || values.length == 0) {
+            throw new IllegalArgumentException("null or empty values (" + values + ")");
+        }
+        this.field = field;
+        this.values = Collections.unmodifiableList(Arrays.asList(values));
+    }
+
+    public String getField() {
+        return field;
+    }
+
+    public List<Object> getValues() {
+        return values;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
+
+        final FieldRoleMapperExpression that = (FieldRoleMapperExpression) o;
+
+        return Objects.equals(this.getField(), that.getField()) && Objects.equals(this.getValues(), that.getValues());
+    }
+
+    @Override
+    public int hashCode() {
+        int result = field.hashCode();
+        result = 31 * result + values.hashCode();
+        return result;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.startObject(NAME);
+        builder.startArray(this.field);
+        for (Object value : values) {
+            builder.value(value);
+        }
+        builder.endArray();
+        builder.endObject();
+        return builder.endObject();
+    }
+
+    public static FieldRoleMapperExpression ofUsername(Object... values) {
+        return ofKeyValues("username", values);
+    }
+
+    public static FieldRoleMapperExpression ofGroups(Object... values) {
+        return ofKeyValues("groups", values);
+    }
+
+    public static FieldRoleMapperExpression ofDN(Object... values) {
+        return ofKeyValues("dn", values);
+    }
+
+    public static FieldRoleMapperExpression ofMetadata(String key, Object... values) {
+        String fieldName = key.startsWith("metadata.") ? key : "metadata." + key;
```

Comment:
I would prefer we require one or the other. That is _always add the "metadata." prefix_ **or** _throw an exception if the prefix is missing_.
Adding it if it doesn't exist feels convenient, but is just a form of leniency that tends to come back and bite us.